daffodil-cli/src/it/scala/org/apache/daffodil/exi/TestEXIEncodeDecode.scala

```diff
@@ -0,0 +1,95 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.saving
+
+import org.junit.Test
+import scala.xml.XML
+import java.nio.file.Paths
+
+import org.apache.daffodil.CLI.Util._
+import org.apache.daffodil.Main.ExitCode
+import org.apache.daffodil.xml.XMLUtils
+
+class TestCLIEncodeDecodeEXI {
+
+  @Test def test_3017_CLI_Encode_Decode_EXI_SA(): Unit = {
+    val schema = path("daffodil-test/src/test/resources/org/apache/daffodil/usertests/Book2.dfdl.xsd")
+    val infosetPath = path("daffodil-test/src/test/resources/org/apache/daffodil/usertests/test_Book2.expected.xml")
+    val infosetXML = XML.loadFile(infosetPath.toFile)
+
+    withTempDir { tempDir =>
+      val tempEXI = Paths.get(tempDir.toString, "temp.exi")
+      val tempXML = Paths.get(tempDir.toString, "temp.xml")
+
+      // Encode infoset to schema aware EXI
+      runCLI(args"exi -s $schema -o $tempEXI") { cli =>
+        cli.sendLine(infosetXML.toString, inputDone = true)
+      } (ExitCode.Success)
+
+      // Decode EXI to XML and compare against original XML infoset
+      runCLI(args"exi -d -s $schema -o $tempXML $tempEXI") { cli =>
+      } (ExitCode.Success)
+
+      val resultNode = XML.loadFile(tempXML.toFile)
+      XMLUtils.compareAndReport(infosetXML, resultNode)
+    }
+  }
+
+  @Test def test_3017_CLI_Encode_Decode_EXI(): Unit = {
+    val inputXML = <person><name>Edward</name><age>42</age></person>
+
+    withTempDir { tempDir =>
+      val tempEXI = Paths.get(tempDir.toString, "temp.exi")
+      val tempXML = Paths.get(tempDir.toString, "temp.xml")
+
+      runCLI(args"exi -o $tempEXI") { cli =>
+        cli.sendLine(inputXML.toString, inputDone = true)
+      } (ExitCode.Success)
+
+      runCLI(args"exi -d -o $tempXML $tempEXI") { cli =>
+      } (ExitCode.Success)
+
+      val resultNode = XML.loadFile(tempXML.toFile)
+      XMLUtils.compareAndReport(inputXML, resultNode)
+    }
+  }
+
+  @Test def test_3017_CLI_EncodeBadFile_EXI(): Unit = {
+    val badXML = path("daffodil-test/src/test/resources/org/apache/daffodil/usertests/Book2.csv")
+
+    runCLI(args"exi $badXML") { cli =>
+      cli.expectErr("Error parsing input XML")
+    } (ExitCode.Failure)
+  }
+
+  @Test def test_3017_CLI_DecodeBadFile_EXI(): Unit = {
+    val badEXI = path("daffodil-test/src/test/resources/org/apache/daffodil/usertests/Book2.csv")
+
+    runCLI(args"exi -d $badEXI") { cli =>
+      cli.expectErr("No valid EXI document")
+    } (ExitCode.Failure)
+  }
+
+  @Test def test_3017_CLI_LoadBadSchema_EXI(): Unit = {
+    val badSchema = path("daffodil-test/src/test/resources/org/apache/daffodil/usertests/Book2.csv")
+
+    runCLI(args"exi -s $badSchema") { cli =>
+      cli.expectErr("Error creating EXI grammar for the supplied schema")
+    } (ExitCode.Failure)
+  }
+}
```

Comment:
Thoughts on two additional tests/checks:
1. Show that it errors if you try to encode an XML file with the `-s` option, but the XML doesn't match the provided schema
1. Show that it errors if you try to decode a schema aware EXI file, but either no `-s` option was given or it was but it doesn't match what the EXI file was encoded with

Reply:
Number 1 will gladly encode a file with a mis-matched shcema and can even save space assuming it finds common elements I think.  It also decodes as expected.

A test for number 2 has been added.